### PR TITLE
Cache results in account witness domain

### PR DESCRIPTION
**Pull request by chimp1984**

### CODE REVIEW

**Comment by chimp1984** [comment]
Here are results from performance tests (based on live network data):

I used a loop for 5 iterations over each offer. For findWitness 10000 iterations.

Version of that PR:
652 offers. 6520000 calls of accountAgeWitnessService.findWitness(offer) took 783 ms  (84% to master)
652 offers. 3260 calls of accountAgeWitnessService.getSignState(accountAgeWitness) took 2063 ms (71% to master)
652 offers. 3260 calls of accountAgeWitnessService.getWitnessSignAge(accountAgeWitness, now) took 4094 ms (78 %to master)
652 offers. 3260 calls of accountAgeWitnessService.getAccountAge(accountAgeWitness, now) took 4094 ms (78% to master)
652 offers. 3260 calls of signedWitnessService.getSignedWitnessSet(accountAgeWitness) took 4095 ms (74% to master) 
652 offers. 3260 calls of signedWitnessService.getWitnessDateList(accountAgeWitness) took 4096 ms (70% to master) 

Master version:
653 offers. 6530000 calls of accountAgeWitnessService.findWitness(offer) took 924 ms 
653 offers. 3265 calls of accountAgeWitnessService.getSignState(accountAgeWitness) took 2873 ms 
653 offers. 3265 calls of accountAgeWitnessService.getWitnessSignAge(accountAgeWitness, now) took 5226 ms 
653 offers. 3265 calls of accountAgeWitnessService.getAccountAge(accountAgeWitness, now) took 5226 ms 
653 offers. 3265 calls of signedWitnessService.getSignedWitnessSet(accountAgeWitness) took 5501 ms 
653 offers. 3265 calls of signedWitnessService.getWitnessDateList(accountAgeWitness) took 5788 ms 

I expected that the findWitness call is more expensive but seems thats not the bottleneck. So the optimization with `accountAgeWitnessCache` is prob. not making much difference. Though that is a pretty trivial one.
The optimization with `getSignedWitnessSetCache` seems to have a bit more impact.

I think it requires further investigations where the bottlenecks are in the tested methods.

Overall that PR brings still about a 20-30% improvement.
There are about 3000-5000 signedWitnessService.getSignedWitnessSet calls with normal user activity after a few minutes which accumulates to about 6-10 sec. cpu time in all.

As its a bit risky domain I have no strong opinion about to merge that PR or close it and delay for a later more broad investigation and improvement. Up to maintainers....

**Comment by chimp1984** [comment]
Seems some Travis issue again: "Could not resolve com.google.api.grpc:proto-google-common-protos:1.12.0." Will force push...

**Comment by chimp1984** [comment]
Travis has issues again... happens quite often recently ;-( Will force push...

File: core/src/main/java/bisq/core/account/sign/SignedWitnessService.java
Comment (sqrrm) [inline core/src/main/java/bisq/core/account/sign/SignedWitnessService.java]: The name is a bit off for a map, just `signedWitnessSetCache` would be more natural.

File: core/src/main/java/bisq/core/account/sign/SignedWitnessService.java
Comment (sqrrm) [inline core/src/main/java/bisq/core/account/sign/SignedWitnessService.java]: This looks redundant, it's removing the same item from the cache again. `hash` is calculated as `new P2PDataStorage.ByteArray(signedWitness.getAccountAgeWitnessHash())` when following the calls to signedWitness.

File: core/src/main/java/bisq/core/account/sign/SignedWitnessService.java
Comment (sqrrm) [inline core/src/main/java/bisq/core/account/sign/SignedWitnessService.java]: This seems redundant, same reason as earlier comment

File: core/src/main/java/bisq/core/account/sign/SignedWitnessService.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/account/sign/SignedWitnessService.java]: Agree...

File: core/src/main/java/bisq/core/account/sign/SignedWitnessService.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/account/sign/SignedWitnessService.java]: I have not looked closely but I took it from the getSignedWitnessSet methods where `signedWitness.getHashAsByteArray()` got compared with `new P2PDataStorage.ByteArray(signedWitness.getAccountAgeWitnessHash())`. Have not looked deeper into the domain if those 2 data are the same actually. If so, yes that second remove is pointless.

**Comment by sqrrm** [comment]
@chimp1984 I'm waiting for your fixes before merging

**Comment by chimp1984** [comment]
@sqrrm I was waiting for your reply to
"I have not looked closely but I took it from the getSignedWitnessSet methods where signedWitness.getHashAsByteArray() got compared with new P2PDataStorage.ByteArray(signedWitness.getAccountAgeWitnessHash()). Have not looked deeper into the domain if those 2 data are the same actually. If so, yes that second remove is pointless."

File: core/src/main/java/bisq/core/account/sign/SignedWitnessService.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/account/sign/SignedWitnessService.java]: I looked up in the code and both fields are different!
```
byte[] data = Utilities.concatenateByteArrays(accountAgeWitnessHash, signature);
data = Utilities.concatenateByteArrays(data, signerPubKey);
hash = Hash.getSha256Ripemd160hash(data);
        
public P2PDataStorage.ByteArray getHashAsByteArray() {
        return new P2PDataStorage.ByteArray(hash);
}
```

**Review by sqrrm — APPROVED** [review]
utACK

@chimp1984 did you want to add anything else to this PR?

**Pull request merged**
